src/rpm_ostree/cli_status.rs

```diff
@@ -36,82 +39,70 @@ lazy_static::lazy_static! {
     )).unwrap();
 }
 
-/// JSON output from `rpm-ostree status --json`
-#[derive(Clone, Debug, Deserialize)]
-pub struct StatusJSON {
-    deployments: Vec<DeploymentJSON>,
-}
-
-/// Partial deployment object (only fields relevant to zincati).
-#[derive(Clone, Debug, Deserialize)]
-#[serde(rename_all = "kebab-case")]
-pub struct DeploymentJSON {
-    booted: bool,
-    base_checksum: Option<String>,
-    #[serde(rename = "base-commit-meta")]
-    base_metadata: BaseCommitMetaJSON,
-    checksum: String,
-    // NOTE(lucab): missing field means "not staged".
-    #[serde(default)]
-    staged: bool,
-    version: String,
-}
-
-/// Metadata from base commit (only fields relevant to zincati).
-#[derive(Clone, Debug, Deserialize)]
-struct BaseCommitMetaJSON {
-    #[serde(rename = "coreos-assembler.basearch")]
-    basearch: String,
-    #[serde(rename = "fedora-coreos.stream")]
-    stream: String,
-}
-
-impl DeploymentJSON {
-    /// Convert into `Release`.
-    pub fn into_release(self) -> Release {
+impl From<&rpmostree_client::Deployment> for Release {
+    fn from(d: &rpmostree_client::Deployment) -> Self {
         Release {
-            checksum: self.base_revision(),
-            version: self.version,
+            checksum: d.base_checksum.clone().unwrap_or_else(|| d.checksum.clone()),
+            version: d.version.clone().unwrap_or_default(),
             age_index: None,
         }
     }
-
-    /// Return the deployment base revision.
-    pub fn base_revision(&self) -> String {
-        self.base_checksum
-            .clone()
-            .unwrap_or_else(|| self.checksum.clone())
-    }
 }
 
 /// Parse base architecture for booted deployment from status object.
-pub fn parse_basearch(status: &StatusJSON) -> Fallible<String> {
-    let json = booted_json(status)?;
-    Ok(json.base_metadata.basearch)
+pub fn parse_basearch(status: &rpmostree_client::Status) -> Fallible<&String> {
+    let d = booted_deployment(status)?;
+    let v = d.base_commit_meta.get(COSA_BASEARCH);
+    if let Some(basearch) = v {
+        match basearch {
+            serde_json::Value::String(basearch) => {
+                ensure!(!basearch.is_empty(), "empty basearch");
+                Ok(basearch)
+            },
+            _ => {
+                failure::bail!("Invalid non-string {} metadata key", COSA_BASEARCH);
+            }
+        }
+    } else {
+        failure::bail!("No {} metadata key", COSA_BASEARCH);
+    }
 }
 
 /// Parse the booted deployment from status object.
-pub fn parse_booted(status: &StatusJSON) -> Fallible<Release> {
-    let json = booted_json(status)?;
-    Ok(json.into_release())
+pub fn parse_booted(status: &rpmostree_client::Status) -> Fallible<Release> {
```

Comment:
Hm, this should probably be `find_booted()` (we're not parsing anything now) and potentially be upstream API too.